test/test_convolutional_layer.h

```diff
@@ -455,48 +455,30 @@ TEST(convolutional, fprop_nnp) {
   convolutional_layer l(5, 5, 3, 1, 2, padding::valid, true, 1, 1,
                         core::backend_t::nnpack);
 
-  // layer::forward_propagation expects tensors, even if we feed only one
-  // input at a time
-  auto create_simple_tensor = [](size_t vector_size) {
-    return tensor_t(1, vec_t(vector_size));
-  };
-
-  // create simple tensors that wrap the payload vectors of the correct size
-  tensor_t in_tensor     = create_simple_tensor(25),
-           out_tensor    = create_simple_tensor(18),
-           a_tensor      = create_simple_tensor(18),
-           weight_tensor = create_simple_tensor(18),
-           bias_tensor   = create_simple_tensor(2);
-
-  // short-hand references to the payload vectors
-  vec_t &in = in_tensor[0], &weight = weight_tensor[0];
+  tensor_buf buf(l, false);
 
   ASSERT_EQ(l.in_shape()[1].size(), size_t(18));  // weight
-
+  // short-hand references to the payload vectors
+  vec_t &in = buf.in_at(0)[0], &weight = buf.in_at(1)[0];
   uniform_rand(in.begin(), in.end(), -1.0, 1.0);
 
-  std::vector<tensor_t *> in_data, out_data;
-  in_data.push_back(&in_tensor);
-  in_data.push_back(&weight_tensor);
-  in_data.push_back(&bias_tensor);
-  out_data.push_back(&out_tensor);
-  out_data.push_back(&a_tensor);
   l.setup(false);
   {
-    l.forward_propagation(in_data, out_data);
+    l.forward_propagation(buf.in_buf(), buf.out_buf());
+
+    vec_t &out = buf.out_at(0)[0];
 
-    vec_t &out = out_tensor[0];
-    for (auto o : out) EXPECT_DOUBLE_EQ(o, tiny_dnn::float_t(0.5));
+    for (auto o : out) EXPECT_DOUBLE_EQ(o, tiny_dnn::float_t(0.0));
   }
 
   // clang-format off
-  weight[0] = 0.3;  weight[1] = 0.1;  weight[2] = 0.2;
-  weight[3] = 0.0;  weight[4] = -0.1; weight[5] = -0.1;
-  weight[6] = 0.05; weight[7] = -0.2; weight[8] = 0.05;
+  weight[0] = 0.3f;  weight[1] = 0.1f;   weight[2] = 0.2f;
+  weight[3] = 0.0f;  weight[4] = -0.1f;  weight[5] = -0.1f;
+  weight[6] = 0.05f; weight[7] = -0.2f;  weight[8] = 0.05f;
 
-  weight[9]  = 0.0; weight[10] = -0.1; weight[11] = 0.1;
-  weight[12] = 0.1; weight[13] = -0.2; weight[14] = 0.3;
-  weight[15] = 0.2; weight[16] = -0.3; weight[17] = 0.2;
+  weight[9]  = 0.0f; weight[10] = -0.1f; weight[11] = 0.1f;
+  weight[12] = 0.1f; weight[13] = -0.2f; weight[14] = 0.3f;
+  weight[15] = 0.2f; weight[16] = -0.3f; weight[17] = 0.2f;
```

Comment:
Maybe you changed these to remove warnings.
when `CNN_USE_DOUBLE` is defined, I prefer older one.